core/lib/state/src/storage_factory/rocksdb_with_memory.rs

```diff
@@ -73,3 +76,110 @@ impl ReadStorage for RocksdbWithMemory {
         }
     }
 }
+
+/// Data structure that keeps a continuous list of batch diffs.
+#[derive(Debug, Default)]
+pub struct BatchDiffs {
+    diffs: VecDeque<BatchDiff>,
+    first_diff_l1_batch_number: Option<L1BatchNumber>,
+}
+
+impl BatchDiffs {
+    /// Creates empty `BatchDiffs`.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Trims diffs that correspond to batches with number less than `trim_up_to`. Does nothing if there are no.
+    pub fn trim_start(&mut self, trim_up_to: L1BatchNumber) {
+        let Some(first_diff_batch_number) = self.first_diff_l1_batch_number else {
+            return;
+        };
+
+        if first_diff_batch_number < trim_up_to {
+            let split_at = (trim_up_to.0 - first_diff_batch_number.0) as usize;
+            self.diffs = self.diffs.split_off(split_at);
+            if self.diffs.is_empty() {
+                self.first_diff_l1_batch_number = None;
+            } else {
+                self.first_diff_l1_batch_number = Some(trim_up_to);
+            }
+        }
+    }
+
+    /// Pushes the diff.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the batch number of the pushed diff is not sequential.
+    pub fn push(&mut self, l1_batch_number: L1BatchNumber, diff: BatchDiff) {
+        if let Some(first_diff_l1_batch_number) = self.first_diff_l1_batch_number {
+            let next_expected_batch_number =
+                first_diff_l1_batch_number + u32::try_from(self.diffs.len()).unwrap();
+            assert_eq!(l1_batch_number, next_expected_batch_number);
+        } else {
+            self.first_diff_l1_batch_number = Some(l1_batch_number);
+        }
+        self.diffs.push_back(diff);
+    }
+
+    /// Returns diffs for `from_l1_batch..=to_l1_batch`.
+    ///
+    /// # Panics
+    ///
+    /// Panics if there is no diff for some element in the range.
+    pub fn range(&self, batch_range: RangeInclusive<L1BatchNumber>) -> Vec<BatchDiff> {
+        let from_l1_batch = *batch_range.start();
+        let to_l1_batch = *batch_range.end();
+
+        let first_diff_number = self.first_diff_l1_batch_number.expect("empty batch_diffs");
+        assert!(from_l1_batch >= first_diff_number);
+        assert!((to_l1_batch.0 as usize) < (first_diff_number.0 as usize) + self.diffs.len());
+
+        let to_skip = from_l1_batch.0 - first_diff_number.0;
+        let to_take = to_l1_batch.0 - from_l1_batch.0 + 1;
+
+        self.diffs
+            .iter()
+            .skip(to_skip as usize)
+            .take(to_take as usize)
```

Comment:
Nit: You can use [`VecDeque::range()`](https://doc.rust-lang.org/std/collections/struct.VecDeque.html#method.range) to be more idiomatic.